mod-source-record-storage-server/src/main/java/org/folio/services/handlers/MarcBibliographicMatchEventHandler.java

```diff
@@ -0,0 +1,200 @@
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Collectors;
+
+import org.apache.commons.lang.StringUtils;
+import org.folio.DataImportEventPayload;
+import org.folio.MatchDetail;
+import org.folio.MatchProfile;
+import org.folio.dao.RecordDao;
+import org.folio.processing.events.services.handler.EventHandler;
+import org.folio.processing.exceptions.EventProcessingException;
+import org.folio.processing.exceptions.MatchingException;
+import org.folio.processing.matching.reader.util.MarcValueReaderUtil;
+import org.folio.processing.value.Value;
+import org.folio.rest.jaxrs.model.EntityType;
+import org.folio.rest.jaxrs.model.Field;
+import org.folio.rest.jaxrs.model.MatchExpression;
+import org.folio.rest.jaxrs.model.ProfileSnapshotWrapper;
+import org.jooq.Condition;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import io.vertx.core.Vertx;
+import io.vertx.core.json.Json;
+import io.vertx.core.json.JsonObject;
+import io.vertx.core.logging.Logger;
+import io.vertx.core.logging.LoggerFactory;
+
+/**
+ * Handler for MARC-MARC matching/not-matching MARC-record by specific fields.
+ */
+@Component
+public class MarcBibliographicMatchEventHandler implements EventHandler {
+
+  private static final Logger LOG = LoggerFactory.getLogger(MarcBibliographicMatchEventHandler.class);
+  private static final String EVENT_HAS_NO_DATA_MSG = "Failed to handle Instance event, cause event payload context does not contain INSTANCE and/or MARC_BIBLIOGRAPHIC data";
+  private static final String MATCHED_ID_MARC_FIELD = "999ffs";
+  private static final String INSTANCE_ID_MARC_FIELD = "999ffi";
+  private static final String INSTANCE_HRID_MARC_FIELD = "001";
+  private static final String FOUND_MULTIPLE_RECORDS_ERROR_MESSAGE = "Found multiple records matching specified conditions";
+  private static final String CANNOT_FIND_RECORDS_ERROR_MESSAGE = "Can`t find records matching specified conditions";
+  private static final String CANNOT_FIND_RECORDS_FOR_MARC_FIELD_ERROR_MESSAGE = "Can`t find records by this MARC-field path: %s";
+
+  private final RecordDao recordDao;
+  private final Vertx vertx;
+
+  @Autowired
+  public MarcBibliographicMatchEventHandler(final RecordDao recordDao, Vertx vertx) {
+    this.recordDao = recordDao;
+    this.vertx = vertx;
+  }
+
+  /**
+   * Handles MARC_BIB_CREATED event.
+   */
+  @Override
+  public CompletableFuture<DataImportEventPayload> handle(DataImportEventPayload dataImportEventPayload) {
+    CompletableFuture<DataImportEventPayload> future = new CompletableFuture<>();
+    HashMap<String, String> context = dataImportEventPayload.getContext();
+    dataImportEventPayload.getEventsChain().add(dataImportEventPayload.getEventType());
+    String recordAsString = context.get(MARC_BIBLIOGRAPHIC.value());
```

Comment:
please add checks for null and look for other places where it might be applicable

Reply:
@KaterynaSenchenko sure! I added all additional check for eventPayload data.